packages/react-test-renderer/src/ReactTestRenderer.js

```diff
@@ -564,40 +560,52 @@ const ReactTestRendererFiber = {
 
   unstable_setNowImplementation: setNowImplementation,
 
-  act(callback: () => void): Thenable {
+  act(callback: () => void | Promise<void>) {
     // note: keep these warning messages in sync with
-    // createNoop.js and ReactTestUtils.js
-    let result = batchedUpdates(callback);
-    if (__DEV__) {
-      if (result !== undefined) {
-        let addendum;
-        if (result !== null && typeof result.then === 'function') {
-          addendum =
-            "\n\nIt looks like you wrote TestRenderer.act(async () => ...) or returned a Promise from it's callback. " +
-            'Putting asynchronous logic inside TestRenderer.act(...) is not supported.\n';
-        } else {
-          addendum = ' You returned: ' + result;
-        }
-        warningWithoutStack(
-          false,
-          'The callback passed to TestRenderer.act(...) function must not return anything.%s',
-          addendum,
-        );
+    // createReactNoop.js and ReactTestUtils.js
+    const result = actedUpdates(callback);
+    if (
+      result !== null &&
+      typeof result === 'object' &&
+      typeof result.then === 'function'
+    ) {
+      let called = false;
+      if (__DEV__) {
+        setTimeout(() => {
+          if (!called) {
```

Comment:
We should only fire this warning if we're in concurrent mode, I think. Otherwise it's a breaking change. So we'll need some way to detect if the callback scheduled concurrent work.

Reply:
Ah, or I guess not because we currently warn if the callback returns a promise. Never mind.

Reply:
Although it does seem like, ideally, we would enforce that you _always_ await `act`. So that microtasks have a chance to fire. Unfortunately that's a breaking change :/

Reply:
just to clarify how we resolved this - I'll verify that jest won't swallow warnings when this is fired 'out of band' in sync tests.